clang/lib/Driver/ToolChains/CommonArgs.cpp

```diff
@@ -1382,6 +1384,10 @@ collectSanitizerRuntimes(const ToolChain &TC, const ArgList &Args,
       StaticRuntimes.push_back("asan_cxx");
   }
 
+  if (!SanArgs.needsSharedRt() && SanArgs.needsRadsanRt() && SanArgs.linkRuntimes()) {
```

Comment:
It appears to me that this added block and the one above can be merged into a single block outside the `if` statement, as:

```cpp
    if (SanArgs.needsRadsanRt() && SanArgs.linkRuntimes()) {
        StaticRuntimes.push_back("radsan");
    }
```

Reply:
Having said that, that observation is based on code logic alone. I'm wondering whether the second one makes any sense. It looks like it doesn't match the pattern of the other sanitizers, which all introduce some other `_cxx` runtime when `SanArgs.needsSharedRt()` is `false`. I wonder whether we originally added this without truly understanding whether it was needed. More investigations needed here!

Reply:
It also seems to be inconsistent with the assertion at line 1516 in `clang/lib/Driver/ToolChains/Darwin.cpp`, which reads:

```cpp
    assert(Sanitize.needsSharedRt() && "Static sanitizer runtimes not supported");
```

... should we just remove this second block and see what happens?

Reply:
This version you see here is a version that works on both linux and darwin, and is minimal.

In my investigation I found that all sanitizers support Static linking:
```
> rg StaticRuntimes
llvm-project/clang/lib/Driver/ToolChains/CommonArgs.cpp
1324:                         SmallVectorImpl<StringRef> &StaticRuntimes,
1325:                         SmallVectorImpl<StringRef> &NonWholeStaticRuntimes,
1326:                         SmallVectorImpl<StringRef> &HelperStaticRuntimes,
1334:        HelperStaticRuntimes.push_back("asan-preinit");
1339:        HelperStaticRuntimes.push_back("memprof-preinit");
1358:        HelperStaticRuntimes.push_back("hwasan-preinit");
1364:    StaticRuntimes.push_back("stats_client");
1368:    HelperStaticRuntimes.push_back("asan_static");
1380:    StaticRuntimes.push_back("asan");
1382:      StaticRuntimes.push_back("asan_cxx");
1386:    StaticRuntimes.push_back("radsan");
1390:    StaticRuntimes.push_back("memprof");
1392:      StaticRuntimes.push_back("memprof_cxx");
1397:      StaticRuntimes.push_back("hwasan_aliases");
1399:        StaticRuntimes.push_back("hwasan_aliases_cxx");
1401:      StaticRuntimes.push_back("hwasan");
1403:        StaticRuntimes.push_back("hwasan_cxx");
1407:    StaticRuntimes.push_back("dfsan");
1409:    StaticRuntimes.push_back("lsan");
1411:    StaticRuntimes.push_back("msan");
1413:      StaticRuntimes.push_back("msan_cxx");
1416:    StaticRuntimes.push_back("tsan");
1418:      StaticRuntimes.push_back("tsan_cxx");
1422:      StaticRuntimes.push_back("ubsan_minimal");
1424:      StaticRuntimes.push_back("ubsan_standalone");
1426:        StaticRuntimes.push_back("ubsan_standalone_cxx");
1430:    NonWholeStaticRuntimes.push_back("safestack");
1435:      StaticRuntimes.push_back("cfi");
1437:      StaticRuntimes.push_back("cfi_diag");
1439:        StaticRuntimes.push_back("ubsan_standalone_cxx");
1443:    NonWholeStaticRuntimes.push_back("stats");
1447:    StaticRuntimes.push_back("scudo_standalone");
1449:      StaticRuntimes.push_back("scudo_standalone_cxx");
1458:  SmallVector<StringRef, 4> SharedRuntimes, StaticRuntimes,
1459:      NonWholeStaticRuntimes, HelperStaticRuntimes, RequiredSymbols;
1461:    collectSanitizerRuntimes(TC, Args, SharedRuntimes, StaticRuntimes,
1462:                             NonWholeStaticRuntimes, HelperStaticRuntimes,
1487:  for (auto RT : HelperStaticRuntimes)
1490:  for (auto RT : StaticRuntimes) {
1494:  for (auto RT : NonWholeStaticRuntimes) {
1523:  return !StaticRuntimes.empty() || !NonWholeStaticRuntimes.empty();

```

But very few do shared. 

```
> rg SharedRuntimes
llvm-project/clang/lib/Driver/ToolChains/CommonArgs.cpp
1323:                         SmallVectorImpl<StringRef> &SharedRuntimes,
1332:      SharedRuntimes.push_back("asan");
1337:      SharedRuntimes.push_back("memprof");
1343:        SharedRuntimes.push_back("ubsan_minimal");
1345:        SharedRuntimes.push_back("ubsan_standalone");
1348:      SharedRuntimes.push_back("scudo_standalone");
1351:      SharedRuntimes.push_back("tsan");
1354:        SharedRuntimes.push_back("hwasan_aliases");
1356:        SharedRuntimes.push_back("hwasan");
1458:  SmallVector<StringRef, 4> SharedRuntimes, StaticRuntimes,
1461:    collectSanitizerRuntimes(TC, Args, SharedRuntimes, StaticRuntimes,
1485:  for (auto RT : SharedRuntimes)
```

When I removed this functionality from ours nothing changed, I was never able to get into the block to even hit an `assert(false)`. 

My proposal. Keep this updated version I have here, that has StaticRuntimes only, but ask in the PR if we should support both.